File: E_n_meb/Source/CreateHH/CheckHH.m
//
//  CheckHH.m
//  E_n_meb
//
//  Created by M on 26.11.15.
//
//

#import "CheckHH.h"
#import "BimodQ.h"
#import "ImMatrix.h"
#import "KoefMatrix.h"
#import "Utils.h"

@implementation CheckHH

+ (BOOL)checkHHElem:(HHElem *)hh degree:(NSInteger)degree
{
    BimodQ *qFrom = [[BimodQ alloc] initForDeg:degree];
    BimodQ *qTo   = [[BimodQ alloc] initForDeg:0];

    if (hh.height != qTo.pij.count) {
        WriteLog(1, "CheckHHElem error! Wrong number of rows");
        return NO;
    }
    if (hh.width != qFrom.pij.count) {
        WriteLog(1, "CheckHHElem error! Wrong number of columns");
        return NO;
    }

    for (NSInteger i = 0; i < hh.height; i++) {
        for (NSInteger j = 0; j < hh.width; j++) {
            if (hh.rows[i][j].isZero) continue;

            Tenzor *t = hh.rows[i][j].content.firstObject.tenzor;
            Vertex *v1 = [[Vertex alloc] initWithI:qFrom.pij[j].n0];

            if (![t.leftComponent.endsWith isEq:v1]) {
                WriteLog(1, "CheckHHElem error! Bad way at pos %d, %d (left ends in %d, must be %d), Diff:",
                         i, j, t.leftComponent.endsWith.number, v1.number);
                Diff *diff = [[Diff alloc] initWithDeg:degree];
                [PrintUtils printMatrixDeg:@"" :diff :degree :degree - 1];
                return NO;
            }

            Vertex *v2 = [[Vertex alloc] initWithI:qTo.pij[i].n0];

            if (![t.leftComponent.startsWith isEq:v2]) {
                WriteLog(1, "CheckHHElem error! Bad way at pos %d, %d (left starts in %d, must be %d), Diff:",
                         i, j, t.leftComponent.startsWith.number, v2.number);
                Diff *diff = [[Diff alloc] initWithDeg:degree];
                [PrintUtils printMatrixDeg:@"" :diff :degree :degree - 1];
                return NO;
            }

            if (![t.rightComponent.startsWith isEq:v2]) {
                WriteLog(1, "CheckHHElem error! Bad way at pos %d, %d (right starts in %d, must be %d), Diff:",
                         i, j, t.rightComponent.startsWith.number, v2.number);
                Diff *diff = [[Diff alloc] initWithDeg:degree];
                [PrintUtils printMatrixDeg:@"" :diff :degree :degree - 1];
                return NO;
            }
        }
    }

    if (![self checkForKer:hh degree:degree]) {
        WriteLog(1, "CheckHHForKer error!");
        return NO;
    }
    if (![self checkForIm:hh degree:degree]) {
        WriteLog(1, "CheckHHForIm error!");
        return NO;
    }

    return YES;
}

+ (BOOL)checkForKer:(HHElem *)hh degree:(NSInteger)degree
{
    Diff *diff = [[Diff alloc] initWithDeg:degree];
    Matrix *multResult = [[Matrix alloc] initWithMult:hh and:diff];
    if (multResult.isZero) {
        return NO;
    }

    NSInteger charK = PathAlg.charK;

    for (NSInteger i = 0; i < multResult.width; i++) {
        float totalKoef = 0;
        Way *columnWay = nil;
        for (NSInteger j = 0; j < multResult.height; j++) {
            Comb *c = multResult.rows[j][i];

            for (TenzorPair *p in c.content) {
                if (!p.koef) continue;
                if (p.tenzor.isZero) continue;

                Way *w = [[Way alloc] initWithWay:p.tenzor.leftComponent];
                Way *wR = [[Way alloc] initFrom:p.tenzor.leftComponent.startsWith.number
                                                 to:p.tenzor.rightComponent.endsWith.number];
                [w compRight:wR];
                [w compRight:p.tenzor.rightComponent];
                if (!w.isZero && !columnWay) {
                    columnWay = [[Way alloc] initWithWay:w];
                }

                if (!w.isZero && ![columnWay isEq:w]) {
                    WriteLog(1, "CheckHHForKer error: various ways (%s and %s) in column %d!", w.htmlStr, columnWay.htmlStr, i);
                    return NO;
                }
                if (!w.isZero) totalKoef += p.koef;
            }
        }

        if ((charK > 0 && ((NSInteger)totalKoef % charK)) || (charK == 0 && totalKoef != 0)) {
            [PrintUtils printMatrix:@"HH" :hh];
            WriteLog(1, "Im");
            ImMatrix *im = [[ImMatrix alloc] initWithDiff:diff];
            [PrintUtils printIm:im deg:degree - 1];
            WriteLog(1, "Error in %d column", i);

            return NO;
        }
    }
    return YES;
}

+ (BOOL)checkForIm:(HHElem *)hh degree:(NSInteger)degree
{
    if (degree == 0) return YES;

    NSInteger s = PathAlg.s;

    Diff *diff = [[Diff alloc] initWithDeg:degree - 1];
    ImMatrix *im = [[ImMatrix alloc] initWithDiff:diff];

    // create line with koefficients
    NSMutableArray<WayPair *> *hhElem = [[NSMutableArray alloc] init];

    for (NSInteger i = 0; i < hh.width; i++) {
        NSInteger j = 0;
        while (j < hh.height) {
            if (!hh.rows[j][i].isZero) break;
            ++j;
        }
        if (j == hh.height) {
            [hhElem addObject:[WayPair pairWithWay:nil koef:0]];
            continue;
        }

        Tenzor *tt = hh.rows[j][i].content.firstObject.tenzor;
        NSInteger hhKoef = (NSInteger)hh.rows[j][i].content.firstObject.koef;

        Way *w = [[Way alloc] initWithWay:tt.leftComponent];
        Way *wR = [[Way alloc] initFrom:tt.leftComponent.startsWith.number to:tt.rightComponent.endsWith.number];
        [w compRight:wR];
        [w compRight:tt.rightComponent];

        j = 0;
        while (j < im.rows.count && (im.rows[j][i].way.isZero || im.rows[j][i].koef == 0))
            j++;

        if (j < im.rows.count) {
            Way* ww = im.rows[j][i].way;

            NSInteger koef = (NSInteger) (hhKoef * ([ww isEq:w] ? 1 : 0));

            [hhElem addObject:[WayPair pairWithWay:ww koef:koef]];

            if (s == 1 && ww.len == 4 && w.len == 0) {
                return YES;
            }
        } else {
            [hhElem addObject:[WayPair pairWithWay:w koef:hhKoef]];
        }
    }

    KoefIntMatrix *k = [[KoefIntMatrix alloc] initWithIm:im];
    NSInteger dimIm = [k rank];

    [im addRow:hhElem];

    if (im.rows.count == 0) {
        WriteLog(1, "Error: bad row size %tu", hhElem.count);
        return NO;
    }

    k = [[KoefIntMatrix alloc] initWithIm:im];
    NSInteger dimImAdd = [k rank];

    if (dimImAdd == dimIm) {
        WriteLog(0, "Im");
        [PrintUtils printIm:im deg:degree - 1];
        WriteLog(0, "dim=%d", dimImAdd);
    }

    return (dimImAdd > dimIm);
}

@end

File: E_n_meb/Source/Utils/CheckHH.swift
//
//  Created by M on 04.02.18.
//

struct CheckHH {
    enum CheckImResult { case failed, inIm, notInIm }

    static func checkHHElem(_ hh: HHElem, degree: Int) -> Bool {
        let qFrom = BimodQ(deg: degree)
        let qTo = BimodQ(deg: 0)
        guard hh.height == qTo.pij.count else {
            OutputFile.writeLog(.error, "CheckHHElem error! Wrong number of rows")
            return false
        }
        guard hh.width == qFrom.pij.count else {
            OutputFile.writeLog(.error, "CheckHHElem error! Wrong number of columns")
            return false
        }
        for i in 0..<hh.height {
            for j in 0..<hh.width {
                guard !hh.rows[i][j].isZero else { continue }
                let t = hh.rows[i][j].content.first!.tenzor

                let v1 = Vertex(i: qFrom.pij[j].n0)
                let v2 = Vertex(i: qTo.pij[i].n0)
                guard t.leftComponent.endsWith.isEq(v1) else {
                    OutputFile.writeLog(.error, "CheckHHElem error! Bad way at pos \(i), \(j) "
                        + "(left ends in \(t.leftComponent.endsWith.number), must be \(v1.number),"
                        + "way is \(Way(from: v2.number, to: v1.number).str))")
                    PrintUtils.printMatrixDeg("Diff tr", Matrix(transpose: Diff(deg: degree)), degree, degree - 1)
                    return false
                }
                guard t.leftComponent.startsWith.isEq(v2) else {
                    OutputFile.writeLog(.error, "CheckHHElem error! Bad way at pos \(i), \(j) "
                        + "(left starts in \(t.leftComponent.startsWith.number), must be \(v2.number),"
                        + "way is \(Way(from: v2.number, to: v1.number).str)")
                    PrintUtils.printMatrixDeg("Diff tr", Matrix(transpose: Diff(deg: degree)), degree, degree - 1)
                    return false
                }
                guard t.rightComponent.startsWith.isEq(v2) else {
                    OutputFile.writeLog(.error, "CheckHHElem error! Bad way at pos \(i), \(j) "
                        + "(right starts in \(t.rightComponent.startsWith.number), must be \(v2.number)")
                    PrintUtils.printMatrixDeg("Diff tr", Matrix(transpose: Diff(deg: degree)), degree, degree - 1)
                    return false
                }
            }
        }
        guard checkForKer(hh, degree: degree) else {
            OutputFile.writeLog(.error, "CheckHHForKer error!")
            return false
        }
        guard checkForIm(hh, degree: degree, shouldBeInIm: false, logError: true) == .notInIm else {
            OutputFile.writeLog(.error, "CheckHHForIm error!")
            return false
        }
        return true
    }

    private static func checkForKer(_ hh: HHElem, degree: Int) -> Bool {
        let diff = Diff(deg: degree)
        let multResult = Matrix(mult: hh, and: diff)
        guard !multResult.isZero else { return false }

        let charK = PathAlg.charK
        for i in 0 ..< multResult.width {
            var totalKoef = 0.0
            var columnWay: Way?
            for j in 0 ..< multResult.height {
                let c = multResult.rows[j][i]
                for p in c.content {
                    guard p.koef != 0 && !p.tenzor.isZero else { continue }
                    let w = Way(way: p.tenzor.leftComponent)
                    let wR = Way(from: p.tenzor.leftComponent.startsWith.number, to: p.tenzor.rightComponent.endsWith.number)
                    w.compRight(wR)
                    w.compRight(p.tenzor.rightComponent)
                    guard !w.isZero else { continue }

                    if let columnWay = columnWay {
                        guard columnWay.isEq(w) else {
                            OutputFile.writeLog(.error, "CheckHHForKer error: various ways (\(w.htmlStr) and \(columnWay.htmlStr) in column \(i)!")
                            return false
                        }
                    } else {
                        columnWay = Way(way: w)
                    }
                    totalKoef += p.koef
                }
            }
            if (charK > 0 && (Int(totalKoef) % charK) != 0) || (charK == 0 && totalKoef != 0) {
                PrintUtils.printMatrix("HH", hh)
                PrintUtils.printIm("Im", ImMatrix(diff: diff))
                OutputFile.writeLog(.error, "Error in \(i) column")
                return false
            }
        }
        return true
    }

    static func checkForIm(_ hh: Matrix, degree: Int, shouldBeInIm: Bool, logError: Bool) -> CheckImResult {
        guard degree > 0 else { return .notInIm }
        let s = PathAlg.s
        let diff = Diff(deg: degree - 1)
        let im = ImMatrix(diff: diff)
        var hhElem: [WayPair] = [] // line with koefficients

        for i in 0 ..< hh.width {
            var hhWay: Way?
            var hhKoef = 0
            for j in 0 ..< hh.height {
                guard !hh.rows[j][i].isZero else { continue }
                if hh.rows[j][i].content.count != 1 {
                    PrintUtils.printMatrix("Error: multiple comb at [\(j),\(i)]: \(hh.rows[j][i].str), hh:", hh, redColumns: [i])
                    return .failed
                }
                let w = ImMatrix.wayForTenzor(hh.rows[j][i].content[0].tenzor) ?? Way()
                if hhWay == nil {
                    hhWay = w
                } else {
                    if (hhWay!.isZero != w.isZero) || (!w.isZero && !w.isEq(hhWay!)) {
                        PrintUtils.printMatrix("Error: different ways at column \(i), hh:", hh, redColumns: [i])
                        return .failed
                    }
                }
                hhKoef += Int(hh.rows[j][i].content[0].koef)
            }
            guard let w = hhWay else { hhElem += [ WayPair() ]; continue }

            for j in 0 ..< im.height {
                if im.rows[j][i].koef != 0, let ww = im.rows[j][i].way, !ww.isZero {
                    if !ww.isEq(w) {
                        PrintUtils.printMatrix("Error: different ways at column \(i), hh:", hh, redColumns: [i])
                        return .failed
                    }
                    if s == 1 && ww.len == 4 && w.len == 0 {
                        return .notInIm
                    }
                }
            }
            hhElem += [ WayPair(way: w, koef: Double(hhKoef)) ]
        }
        let dimIm = KoefIntMatrix(im: im).rank
        im.addRow(hhElem)
        if im.rows.count == 0 {
            OutputFile.writeLog(.error, "Error: bad row size \(hhElem.count)")
            return .failed
        }
        let dimImAdd = KoefIntMatrix(im: im).rank
        if logError && ((dimImAdd == dimIm && !shouldBeInIm) || (dimImAdd > dimIm && shouldBeInIm)) {
            PrintUtils.printMatrix("HH", hh)
            PrintUtils.printIm("Im", im)
            OutputFile.writeLog(.normal, "dim=\(dimImAdd)")
        }
        return dimImAdd == dimIm ? .inIm : .notInIm
    }
}
